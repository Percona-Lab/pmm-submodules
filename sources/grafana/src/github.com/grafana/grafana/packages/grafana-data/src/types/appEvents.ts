import { eventFactory } from './utils';

export interface AppEvent<T> {
  readonly name: string;
  payload?: T;
}

export type AlertPayload = [string, string?];
export type AlertErrorPayload = [string, (string | Error)?];

export const alertSuccess = eventFactory<AlertPayload>('alert-success');
export const alertWarning = eventFactory<AlertPayload>('alert-warning');
export const alertError = eventFactory<AlertErrorPayload>('alert-error');
