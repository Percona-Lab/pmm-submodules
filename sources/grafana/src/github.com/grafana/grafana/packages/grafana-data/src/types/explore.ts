import { RawTimeRange } from './time';

/** @internal */
export interface ExploreUrlState {
  datasource: string;
  queries: any[]; // Should be a DataQuery, but we're going to strip refIds, so typing makes less sense
  range: RawTimeRange;
  originPanelId?: number;
  context?: string;
}
