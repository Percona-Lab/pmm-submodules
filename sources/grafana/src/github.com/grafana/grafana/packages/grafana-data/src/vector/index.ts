export * from './AppendedVectors';
export * from './ArrayVector';
export * from './CircularVector';
export * from './ConstantVector';
export * from './BinaryOperationVector';
export * from './SortedVector';
export * from './FormattedVector';

export { vectorator } from './FunctionalVector';
