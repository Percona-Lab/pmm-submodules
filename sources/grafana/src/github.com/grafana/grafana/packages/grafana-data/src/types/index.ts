export * from './data';
export * from './dataFrame';
export * from './dataLink';
export * from './annotations';
export * from './logs';
export * from './navModel';
export * from './select';
export * from './time';
export * from './thresholds';
export * from './utils';
export * from './valueMapping';
export * from './displayValue';
export * from './graph';
export * from './ScopedVars';
export * from './transformations';
export * from './fieldOverrides';
export * from './vector';
export * from './app';
export * from './datasource';
export * from './panel';
export * from './plugin';
export * from './thresholds';
export * from './templateVars';
export * from './fieldColor';
export * from './theme';
export * from './orgs';
export * from './flot';
export * from './trace';
export * from './explore';
export * from './live';

import * as AppEvents from './appEvents';
import { AppEvent } from './appEvents';
export { AppEvent, AppEvents };

import * as PanelEvents from './panelEvents';
export { PanelEvents };
export { GrafanaConfig, BuildInfo, FeatureToggles, LicenseInfo } from './config';
