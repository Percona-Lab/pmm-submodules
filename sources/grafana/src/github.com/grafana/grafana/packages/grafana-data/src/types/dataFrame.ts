import { ThresholdsConfig } from './thresholds';
import { ValueMapping } from './valueMapping';
import { QueryResultBase, Labels, NullValueMode } from './data';
import { DisplayProcessor, DisplayValue } from './displayValue';
import { DataLink, LinkModel } from './dataLink';
import { Vector } from './vector';
import { FieldColor } from './fieldColor';
import { ScopedVars } from './ScopedVars';

export enum FieldType {
  time = 'time', // or date
  number = 'number',
  string = 'string',
  boolean = 'boolean',
  // Used to detect that the value is some kind of trace data to help with the visualisation and processing.
  trace = 'trace',
  other = 'other', // Object, Array, etc
}

/**
 * Every property is optional
 *
 * Plugins may extend this with additional properties. Something like series overrides
 */
export interface FieldConfig<TOptions extends object = any> {
  /**
   * The display value for this field.  This supports template variables blank is auto
   */
  displayName?: string;

  /**
   * This can be used by data sources that return and explicit naming structure for values and labels
   * When this property is configured, this value is used rather than the default naming strategy.
   */
  displayNameFromDS?: string;

  /**
   * True if data source field supports ad-hoc filters
   */
  filterable?: boolean;

  // Numeric Options
  unit?: string;
  decimals?: number | null; // Significant digits (for display)
  min?: number | null;
  max?: number | null;

  // Convert input values into a display string
  mappings?: ValueMapping[];

  // Map numeric values to states
  thresholds?: ThresholdsConfig;

  // Map values to a display color
  color?: FieldColor;

  // Used when reducing field values
  nullValueMode?: NullValueMode;

  // The behavior when clicking on a result
  links?: DataLink[];

  // Alternative to empty string
  noValue?: string;

  // Panel Specific Values
  custom?: TOptions;
}

export interface ValueLinkConfig {
  /**
   * Result of field reduction
   */
  calculatedValue?: DisplayValue;
  /**
   * Index of the value row within Field. Should be provided only when value is not a result of a reduction
   */
  valueRowIndex?: number;
}

export interface Field<T = any, V = Vector<T>> {
  /**
   * Name of the field (column)
   */
  name: string;
  /**
   *  Field value type (string, number, etc)
   */
  type: FieldType;
  /**
   *  Meta info about how field and how to display it
   */
  config: FieldConfig;
  values: V; // The raw field values
  labels?: Labels;

  /**
   * Cached values with appropriate display and id values
   */
  state?: FieldState | null;

  /**
   * Convert text to the field value
   */
  parse?: (value: any) => T;

  /**
   * Convert a value for display
   */
  display?: DisplayProcessor;

  /**
   * Get value data links with variables interpolated
   */
  getLinks?: (config: ValueLinkConfig) => Array<LinkModel<Field>>;
}

export interface FieldState {
  /**
   * An appropriate name for the field (does not include frame info)
   */
  displayName?: string | null;

  /**
   * Cache of reduced values
   */
  calcs?: FieldCalcs;

  /**
   * Appropriate values for templating
   */
  scopedVars?: ScopedVars;

  /**
   * Series index is index for this field in a larger data set that can span multiple DataFrames
   * Useful for assigning color to series by looking up a color in a palette using this index
   */
  seriesIndex?: number;
}

export interface DataFrame extends QueryResultBase {
  name?: string;
  fields: Field[]; // All fields of equal length

  // The number of rows
  length: number;
}

/**
 * Like a field, but properties are optional and values may be a simple array
 */
export interface FieldDTO<T = any> {
  name: string; // The column name
  type?: FieldType;
  config?: FieldConfig;
  values?: Vector<T> | T[]; // toJSON will always be T[], input could be either
  labels?: Labels;
}

/**
 * Like a DataFrame, but fields may be a FieldDTO
 */
export interface DataFrameDTO extends QueryResultBase {
  name?: string;
  fields: Array<FieldDTO | Field>;
}

export interface FieldCalcs extends Record<string, any> {}

export const TIME_SERIES_VALUE_FIELD_NAME = 'Value';
export const TIME_SERIES_TIME_FIELD_NAME = 'Time';
export const TIME_SERIES_METRIC_FIELD_NAME = 'Metric';
