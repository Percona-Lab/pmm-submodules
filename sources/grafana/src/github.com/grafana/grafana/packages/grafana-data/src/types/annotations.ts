import { Observable } from 'rxjs';
import { ComponentType } from 'react';

import { DataQuery, QueryEditorProps } from './datasource';
import { DataFrame } from './dataFrame';

/**
 * This JSON object is stored in the dashboard json model.
 */
export interface AnnotationQuery<TQuery extends DataQuery = DataQuery> {
  datasource: string;
  enable: boolean;
  name: string;
  iconColor: string;
  hide?: boolean;

  // Standard datasource query
  target?: TQuery;

  // Convert a dataframe to an AnnotationEvent
  mappings?: AnnotationEventMappings;
}

export interface AnnotationEvent {
  id?: string;
  annotation?: any;
  dashboardId?: number;
  panelId?: number;
  userId?: number;
  login?: string;
  email?: string;
  avatarUrl?: string;
  time?: number;
  timeEnd?: number;
  isRegion?: boolean;
  title?: string;
  text?: string;
  type?: string;
  tags?: string[];
  color?: string;

  // Currently used to merge annotations from alerts and dashboard
  source?: any; // source.type === 'dashboard'
}

/**
 * @alpha -- any value other than `field` is experimental
 */
export enum AnnotationEventFieldSource {
  Field = 'field', // Default -- find the value with a matching key
  Text = 'text', // Write a constant string into the value
  Skip = 'skip', // Do not include the field
}

export interface AnnotationEventFieldMapping {
  source?: AnnotationEventFieldSource; // defaults to 'field'
  value?: string;
  regex?: string;
}

export type AnnotationEventMappings = Partial<Record<keyof AnnotationEvent, AnnotationEventFieldMapping>>;

/**
 * Since Grafana 7.2
 *
 * This offers a generic approach to annotation processing
 */
export interface AnnotationSupport<TQuery extends DataQuery = DataQuery, TAnno = AnnotationQuery<TQuery>> {
  /**
   * This hook lets you manipulate any existing stored values before running them though the processor.
   * This is particularly helpful when dealing with migrating old formats.  ie query as a string vs object
   */
  prepareAnnotation?(json: any): TAnno;

  /**
   * Convert the stored JSON model to a standard datasource query object.
   * This query will be executed in the datasource and the results converted into events.
   * Returning an undefined result will quietly skip query execution
   */
  prepareQuery?(anno: TAnno): TQuery | undefined;

  /**
   * When the standard frame > event processing is insufficient, this allows explicit control of the mappings
   */
  processEvents?(anno: TAnno, data: DataFrame[]): Observable<AnnotationEvent[] | undefined>;

  /**
   * Specify a custom QueryEditor for the annotation page.  If not specified, the standard one will be used
   */
  QueryEditor?: ComponentType<QueryEditorProps<any, TQuery>>;
}
