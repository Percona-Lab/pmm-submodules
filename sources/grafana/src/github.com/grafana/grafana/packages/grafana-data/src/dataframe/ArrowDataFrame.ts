import { DataFrame, FieldType, Field, Vector } from '../types';
import { FunctionalVector } from '../vector/FunctionalVector';

import {
  Table,
  ArrowType,
  Builder,
  Vector as ArrowVector,
  Float64,
  DataType,
  Utf8,
  TimestampMillisecond,
  Bool,
  Column,
} from 'apache-arrow';
import { getFieldDisplayName } from '../field';

export interface ArrowDataFrame extends DataFrame {
  table: Table;
}

export function base64StringToArrowTable(text: string): Table {
  const b64 = atob(text);
  const arr = Uint8Array.from(b64, c => {
    return c.charCodeAt(0);
  });
  return Table.from(arr);
}

function valueOrUndefined(val?: string) {
  return val ? val : undefined;
}

function parseOptionalMeta(str?: string): any {
  if (str && str.length && str !== '{}') {
    try {
      return JSON.parse(str);
    } catch (err) {
      console.warn('Error reading JSON from arrow metadata: ', str);
    }
  }
  return undefined;
}

export function arrowTableToDataFrame(table: Table): ArrowDataFrame {
  const fields: Field[] = [];

  for (let i = 0; i < table.numCols; i++) {
    const col = table.getColumnAt(i);
    if (col) {
      const schema = table.schema.fields[i];
      let type = FieldType.other;
      let values: Vector<any> = col;
      switch ((schema.typeId as unknown) as ArrowType) {
        case ArrowType.Decimal:
        case ArrowType.FloatingPoint: {
          type = FieldType.number;
          break;
        }
        case ArrowType.Int: {
          type = FieldType.number;
          values = new NumberColumn(col); // Cast to number
          break;
        }
        case ArrowType.Bool: {
          type = FieldType.boolean;
          break;
        }
        case ArrowType.Timestamp: {
          type = FieldType.time;
          break;
        }
        case ArrowType.Utf8: {
          type = FieldType.string;
          break;
        }
        default:
          console.error('UNKNOWN Type:', schema);
      }

      fields.push({
        name: col.name,
        type,
        values,
        config: parseOptionalMeta(col.metadata.get('config')) || {},
        labels: parseOptionalMeta(col.metadata.get('labels')),
      });
    }
  }
  const meta = table.schema.metadata;
  return {
    fields,
    length: table.length,
    refId: valueOrUndefined(meta.get('refId')),
    name: valueOrUndefined(meta.get('name')),
    meta: parseOptionalMeta(meta.get('meta')),
    table,
  };
}

function toArrowVector(field: Field): ArrowVector {
  // OR: Float64Vector.from([1, 2, 3]));

  let type: DataType;
  if (field.type === FieldType.number) {
    type = new Float64();
  } else if (field.type === FieldType.time) {
    type = new TimestampMillisecond();
  } else if (field.type === FieldType.boolean) {
    type = new Bool();
  } else if (field.type === FieldType.string) {
    type = new Utf8();
  } else {
    type = new Utf8();
  }
  const builder = Builder.new({ type, nullValues: [null] });
  field.values.toArray().forEach(builder.append.bind(builder));
  return builder.finish().toVector();
}

/**
 * @param keepOriginalNames by default, the exported Table will get names that match the
 * display within grafana.  This typically includes any labels defined in the metadata.
 *
 * When using this function to round-trip data, be sure to set `keepOriginalNames=true`
 */
export function grafanaDataFrameToArrowTable(data: DataFrame, keepOriginalNames?: boolean): Table {
  // Return the original table
  let table = (data as any).table;
  if (table instanceof Table && table.numCols === data.fields.length) {
    if (!keepOriginalNames) {
      table = updateArrowTableNames(table, data);
    }
    if (table) {
      return table as Table;
    }
  }

  table = Table.new(
    data.fields.map((field, index) => {
      let name = field.name;
      // when used directly as an arrow table the name should match the arrow schema
      if (!keepOriginalNames) {
        name = getFieldDisplayName(field, data);
      }
      const column = Column.new(name, toArrowVector(field));
      if (field.labels) {
        column.metadata.set('labels', JSON.stringify(field.labels));
      }
      if (field.config) {
        column.metadata.set('config', JSON.stringify(field.config));
      }
      return column;
    })
  );
  const metadata = table.schema.metadata;
  if (data.name) {
    metadata.set('name', data.name);
  }
  if (data.refId) {
    metadata.set('refId', data.refId);
  }
  if (data.meta) {
    metadata.set('meta', JSON.stringify(data.meta));
  }
  return table;
}

function updateArrowTableNames(table: Table, frame: DataFrame): Table | undefined {
  const cols: Column[] = [];
  for (let i = 0; i < table.numCols; i++) {
    const col = table.getColumnAt(i);
    if (!col) {
      return undefined;
    }
    const name = getFieldDisplayName(frame.fields[i], frame);
    cols.push(Column.new(col.field.clone({ name: name }), ...col.chunks));
  }
  return Table.new(cols);
}

class NumberColumn extends FunctionalVector<number> {
  constructor(private col: Column) {
    super();
  }

  get length() {
    return this.col.length;
  }

  get(index: number): number {
    const v = this.col.get(index);
    if (v === null || isNaN(v)) {
      return v;
    }

    // The conversion operations are always silent, never give errors,
    // but if the bigint is too huge and wonâ€™t fit the number type,
    // then extra bits will be cut off, so we should be careful doing such conversion.
    // See https://javascript.info/bigint
    return Number(v);
  }
}
