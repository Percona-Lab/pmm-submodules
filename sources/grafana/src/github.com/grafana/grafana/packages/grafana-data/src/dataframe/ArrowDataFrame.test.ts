import fs from 'fs';
import path from 'path';

import { grafanaDataFrameToArrowTable, arrowTableToDataFrame } from './ArrowDataFrame';
import { toDataFrameDTO, toDataFrame } from './processDataFrame';
import { FieldType, DataFrame } from '../types';
import { Table } from 'apache-arrow';
import { ArrayVector } from '../vector';

describe('Read/Write arrow Table to DataFrame', () => {
  test('should parse output with dataframe', () => {
    const frame = toDataFrame({
      name: 'Hello',
      refId: 'XYZ',
      meta: {
        aaa: 'xyz',
        anything: 'xxx',
      },
      fields: [
        { name: 'time', config: {}, type: FieldType.time, values: [1, 2, 3] },
        { name: 'value', config: { min: 0, max: 50, unit: 'something' }, type: FieldType.number, values: [1, 2, 3] },
        { name: 'str', config: {}, type: FieldType.string, values: ['a', 'b', 'c'] },
      ],
    });

    const table = grafanaDataFrameToArrowTable(frame);
    expect(table.length).toEqual(frame.length);

    // Now back to DataFrame
    const before = JSON.stringify(toDataFrameDTO(frame), null, 2);
    const after = JSON.stringify(toDataFrameDTO(arrowTableToDataFrame(table)), null, 2);
    expect(after).toEqual(before);
  });

  test('should support duplicate field names', () => {
    const frame = toDataFrame({
      name: 'Hello',
      refId: 'XYZ',
      fields: [
        { name: 'time', config: {}, type: FieldType.time, values: [1, 2, 3] },
        { name: 'a', values: [1, 2, 3] },
        { name: 'a', values: ['a', 'b', 'c'] },
      ],
    });

    const table = grafanaDataFrameToArrowTable(frame, true);
    expect(table.length).toEqual(frame.length);

    // Now back to DataFrame
    const before = JSON.stringify(toDataFrameDTO(frame), null, 2);
    const after = JSON.stringify(toDataFrameDTO(arrowTableToDataFrame(table)), null, 2);
    expect(after).toEqual(before);
  });

  test('should read all types', () => {
    const fullpath = path.resolve(__dirname, './__snapshots__/all_types.golden.arrow');
    const arrow = fs.readFileSync(fullpath);
    const table = Table.from([arrow]);
    const frame = arrowTableToDataFrame(table);
    expect(toDataFrameDTO(frame)).toMatchSnapshot();
  });

  test('Export arrow table names', () => {
    const simple: DataFrame = {
      name: 'hello',
      fields: [
        {
          name: 'fname',
          labels: {
            a: 'AAA',
            b: 'BBB',
          },
          config: {},
          type: FieldType.number,
          values: new ArrayVector([1, 2]),
        },
      ],
      length: 2,
    };
    const t1 = grafanaDataFrameToArrowTable(simple);
    const t2 = grafanaDataFrameToArrowTable(simple, true);
    expect(t1.getColumnAt(0)?.name).toEqual('fname {a="AAA", b="BBB"}');
    expect(t2.getColumnAt(0)?.name).toEqual('fname');
  });
});
