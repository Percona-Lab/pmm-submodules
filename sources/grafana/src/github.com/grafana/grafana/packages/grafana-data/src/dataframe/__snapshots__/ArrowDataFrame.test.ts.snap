// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Read/Write arrow Table to DataFrame should read all types 1`] = `
Object {
  "fields": Array [
    Object {
      "config": Object {},
      "labels": Object {
        "aLabelKey": "aLabelValue",
      },
      "name": "string_values",
      "type": "string",
      "values": Array [
        "Go Min",
        "JS Min (for >= 64)",
        "0 / nil / misc",
        "JS Max (for >= 64)",
        "Go Max",
      ],
    },
    Object {
      "config": Object {
        "decimals": 2,
        "displayName": "Grafana ‚ù§Ô∏è (Previous should be heart emoji) ü¶• (Previous should be sloth emoji)",
        "filterable": false,
        "links": Array [
          Object {
            "targetBlank": true,
            "title": "Donate - The Sloth Conservation Foundation",
            "url": "https://slothconservation.com/how-to-help/donate/",
          },
        ],
        "max": null,
        "min": null,
        "noValue": "üò§",
        "nullValueMode": "null",
      },
      "labels": Object {
        "aLabelKey": "aLabelValue",
        "bLabelKey": "bLabelValue",
      },
      "name": "nullable_string_values",
      "type": "string",
      "values": Array [
        "Grafana",
        "‚ù§Ô∏è",
        null,
        "ü¶•",
        "update your unicode/font if no sloth, is 2019.",
      ],
    },
    Object {
      "config": Object {
        "max": 1,
        "min": 0,
      },
      "labels": undefined,
      "name": "int8_values",
      "type": "number",
      "values": Array [
        -128,
        -128,
        0,
        127,
        127,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_int8_values",
      "type": "number",
      "values": Array [
        -128,
        -128,
        null,
        127,
        127,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "int16_values",
      "type": "number",
      "values": Array [
        -32768,
        -32768,
        0,
        32767,
        32767,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_int16_values",
      "type": "number",
      "values": Array [
        -32768,
        -32768,
        null,
        32767,
        32767,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "int32_values",
      "type": "number",
      "values": Array [
        -2147483648,
        -2147483648,
        1,
        2147483647,
        2147483647,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_int32_values",
      "type": "number",
      "values": Array [
        -2147483648,
        -2147483648,
        null,
        2147483647,
        2147483647,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "int64_values",
      "type": "number",
      "values": Array [
        -9223372036854776000,
        -9007199254740991,
        1,
        9007199254740991,
        9223372036854776000,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_int64_values",
      "type": "number",
      "values": Array [
        -9223372036854776000,
        -9007199254740991,
        null,
        9007199254740991,
        9223372036854776000,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "uint8_values",
      "type": "number",
      "values": Array [
        0,
        0,
        1,
        255,
        255,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_uint8_values",
      "type": "number",
      "values": Array [
        0,
        0,
        null,
        255,
        255,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "uint16_values",
      "type": "number",
      "values": Array [
        0,
        0,
        1,
        65535,
        65535,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_uint16_values",
      "type": "number",
      "values": Array [
        0,
        0,
        null,
        65535,
        65535,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "uint32_values",
      "type": "number",
      "values": Array [
        0,
        0,
        1,
        4294967295,
        4294967295,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_uint32_values",
      "type": "number",
      "values": Array [
        0,
        0,
        null,
        4294967295,
        4294967295,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "uint64_values",
      "type": "number",
      "values": Array [
        0,
        0,
        1,
        9007199254740991,
        18446744073709552000,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_uint64_values",
      "type": "number",
      "values": Array [
        0,
        0,
        null,
        9007199254740991,
        18446744073709552000,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "float32_values",
      "type": "number",
      "values": Array [
        1.401298464324817e-45,
        1.401298464324817e-45,
        1,
        3.4028234663852886e+38,
        3.4028234663852886e+38,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_float32_values",
      "type": "number",
      "values": Array [
        1.401298464324817e-45,
        1.401298464324817e-45,
        null,
        3.4028234663852886e+38,
        3.4028234663852886e+38,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "float64_values",
      "type": "number",
      "values": Array [
        5e-324,
        -9007199254740991,
        1,
        9007199254740991,
        1.7976931348623157e+308,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_float64_values",
      "type": "number",
      "values": Array [
        5e-324,
        -9007199254740991,
        null,
        1.7976931348623157e+308,
        9007199254740991,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "bool_values",
      "type": "boolean",
      "values": Array [
        true,
        false,
        true,
        true,
        false,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_bool_values",
      "type": "boolean",
      "values": Array [
        true,
        false,
        null,
        true,
        false,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "timestamps",
      "type": "time",
      "values": Array [
        0,
        1568039445000,
        1568039450000,
        9007199254.740992,
        9223372036854.775,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "timestamps",
      "type": "time",
      "values": Array [
        0,
        1568039445000,
        1568039450000,
        9007199254.740992,
        9223372036854.775,
      ],
    },
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "nullable_timestamps",
      "type": "time",
      "values": Array [
        0,
        1568039445000,
        null,
        9007199254.740992,
        9223372036854.775,
      ],
    },
  ],
  "meta": Object {
    "custom": Object {
      "Hi": "there",
    },
  },
  "name": "many_types",
  "refId": "A",
}
`;
